### PR TITLE
Enforce a certain number of tokens on redemption

### LINKED ISSUE #99: Verify that the number of tokens provided in a redemption request matches expectations.
Otherwise, users can get either more or less ZKAPs than they should.

### PR DESCRIPTION
**Pull request by exarkun**
Fixes #99

### CODE REVIEW

**Review by hacklschorsch — APPROVED** [review]
Thanks for asking for a Review @exarkun. All I can say so far is I `[✓] Viewed` all files, and haven't found any errors obvious to me. That doesn't mean a whole lot currently I am afraid though, so beware :)

**Comment by exarkun** [comment]
> Thanks for asking for a Review @exarkun. All I can say so far is I `[✓] Viewed` all files, and haven't found any errors obvious to me. That doesn't mean a whole lot currently I am afraid though, so beware :)

Thanks @hacklschorsch !

File: src/PaymentServer/Redemption.hs [hunk 1]
```diff
@@ -190,40 +203,74 @@ retry op =
...
+redeem (RedemptionConfig numGroups tokensPerVoucher issue) database (Redeem voucher tokens counter) =
+  if counter < 0 || counter >= numGroups then
+    throwError $ jsonErr err400 (CounterOutOfBounds 0 numGroups counter)
+  else
+    case tokenCountForGroup numGroups tokensPerVoucher (fromIntegral counter) of
+      Nothing ->
+        throwError $ jsonErr err500 $ OtherFailure "invalid redemption counter"
+      Just allowedTokenCount ->
+        if allowedTokenCount /= length tokens
+        then throwError $ jsonErr err400 $ OtherFailure "wrong number of tokens"
```
Comment (tomprince) [inline src/PaymentServer/Redemption.hs:242]: Should this have a structured error message, with the expected number of tokens?

File: src/PaymentServer/Redemption.hs [hunk 2]
```diff
@@ -85,12 +91,23 @@ import PaymentServer.Issuer
...
+  -- ^ An issuer which can actually issue signatures during a redemption.
+  }
+
 data Result
   = Unpaid -- ^ A voucher has not been paid for.
   | DoubleSpend -- ^ A voucher has already been redeemed.
   | OtherFailure Text -- ^ Some other unrecognized failure mode.
   -- | Given counter was not in the expected range
-  | CounterOutOfBounds Integer Integer Integer
+  | CounterOutOfBounds Int Int Int
```
Comment (tomprince) [inline src/PaymentServer/Redemption.hs:110]: Not really related to this PR, but I noticed it while trying to understand it:

I don't think any of these values need to be a part of `Result`, as we never return them as values of type `Result` (but rather encode them as part of `ServerError`). I'm not sure how it should be structured, though [this](https://docs.servant.dev/en/stable/cookbook/uverb/UVerb.html) has some discussion of various possibilities, including some packages that help with it.

**Review by tomprince — COMMENTED** [review]
This looks good.

File: src/PaymentServer/Redemption.hs [hunk 3]
```diff
@@ -190,40 +203,74 @@ retry op =
         Left NotPaid -> return True
         _ -> return False
 
+
+-- | Compute the number of tokens which should be present in a certain
+-- redemption group.
+tokenCountForGroup
+  :: Int
+  -- ^ The total number of redemption groups.
```
[the remark this answers is not in the document]
Comment (exarkun) [reply src/PaymentServer/Redemption.hs:211]: Filed https://github.com/PrivateStorageio/ZKAPAuthorizer/issues/227

File: src/PaymentServer/Redemption.hs [hunk 4]
```diff
@@ -190,40 +203,74 @@ retry op =
...
-      redeem :: IO (Either RedeemError Bool)
-      redeem = redeemVoucherWithCounter database voucher fingerprint counter
+redeem :: VoucherDatabase d => RedemptionConfig -> d -> Redeem -> Handler Result
+redeem (RedemptionConfig numGroups tokensPerVoucher issue) database (Redeem voucher tokens counter) =
+  if counter < 0 || counter >= numGroups then
+    throwError $ jsonErr err400 (CounterOutOfBounds 0 numGroups counter)
+  else
+    case tokenCountForGroup numGroups tokensPerVoucher (fromIntegral counter) of
+      Nothing ->
+        throwError $ jsonErr err500 $ OtherFailure "invalid redemption counter"
```
[the remark this answers is not in the document]
Comment (exarkun) [reply src/PaymentServer/Redemption.hs:239]: I agree this code could use some refactoring to remove the duplication / dead error handling paths.

File: src/PaymentServer/Redemption.hs (hunk 1 above)
Comment (exarkun) [reply src/PaymentServer/Redemption.hs:242]: Maybe.  I guess that would be easier to answer if we had a UX for handling any of these failure conditions.  Right now what happens is ZKAPAuthorizer notes the error and usually tries exactly the same thing again a little later and no software exists to surface these error conditions to the user.

**Pull request merged**